File: 10. Coin Change (Minimum Coins).cpp
// 10. Coin Change (Minimum Coins)
// Topic: Coin Change (Minimum Coins)
// Description: Given coins[] and amount, return minimum number of coins required using
// recursion+memo. Input: amount k, n (#coins), then coin values Output: minimum coins or -1 if not
// possible

#include <bits/stdc++.h>
using namespace std;
using ll = long long;

// Memoization starts here: vector<int> memo for amounts
int minCoins(int amount, const vector<int>& coins, vector<int>& memo)
{
    if (amount == 0)
        return 0;
    if (amount < 0)
        return INT_MAX / 2;
    if (memo[amount] != -1)
        return memo[amount];
    int res = INT_MAX / 2;
    for (int c : coins)
    {
        int sub = minCoins(amount - c, coins, memo);
        if (sub + 1 < res)
            res = sub + 1;
    }
    memo[amount] = res;
    return res;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int amount, n;
    if (!(cin >> amount >> n))
        return 0;
    vector<int> coins(n);
    for (int i = 0; i < n; ++i) cin >> coins[i];
    vector<int> memo(amount + 1, -1);
    int ans = minCoins(amount, coins, memo);
    if (ans >= INT_MAX / 2)
        cout << -1 << '\n';
    else
        cout << ans << '\n';
}

// Memoization process started in minCoins where memo[amount] caches the minimum coins for that
// amount. Time Complexity: O(n * amount) where n is #coins Space Complexity: O(amount)

File: 10. Coin Change (Minimum Coins).py
# 10. Coin Change (Minimum Coins)
# Topic: Coin Change (Minimum Coins)
# Description: Find the minimum number of coins required to make given amount using recursion + memoization.
# Input: amount k, n (#coins), then n coin values
# Output: minimum coins or -1 if not possible

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    amount = int(next(it))
    n = int(next(it))
    coins = [int(next(it)) for _ in range(n)]

    @lru_cache(maxsize=None)
    def min_coins(rem: int) -> int:
        if rem == 0:
            return 0
        if rem < 0:
            return 10**9
        best = 10**9
        for c in coins:
            best = min(best, 1 + min_coins(rem - c))
        return best

    ans = min_coins(amount)
    print(-1 if ans >= 10**9 else ans)

# Memoization process started at min_coins decorated with lru_cache.
# Time Complexity: O(n * amount)
# Space Complexity: O(amount)